packages/react-reconciler/src/ReactChildFiber.js

```diff
@@ -157,7 +156,17 @@ function coerceRef(
         return current.ref;
       }
       const ref = function(value) {
-        const refs = inst.refs === emptyObject ? (inst.refs = {}) : inst.refs;
+        let refs;
+        // This flag tells us if we need to initialize `this.refs`.
+        if (inst.__reactInternalHasLegacyRefs) {
```

Comment:
Small not-even-worth-mentioning nit: since this isn't always a boolean, might be better to explicitly compare to `undefined` instead, like we do for other optional types.